### PR TITLE
Joplin

### LINKED ISSUE #19585: [Package request]: Joplin, or Elephant: an open Evernote alternative(s)
Either https://github.com/laurent22/joplin

or 

https://github.com/jusu/Elephant

(or both :-)
Thanks,

### PR DESCRIPTION
**Pull request by fosslinux**
This PR adds the `joplin-desktop` and `joplin-cli` packages, containing the GUI and CLI versions of Joplin respectively. They are not distributed in the one package as they can and should be installed separately, they are separate applications.

Closes #19585

### CODE REVIEW

**Comment by fosslinux** [comment]
I cannot reproduce this travis failure....

**Comment by ericonr** [comment]
> sh: electron-builder: command not found

weird...

**Comment by ndowens** [comment]
> > sh: electron-builder: command not found
> 
> weird...

I don't know if this would help, but searching I found https://github.com/electron-userland/electron-builder

**Comment by fosslinux** [comment]
... yes, I found that too... and it means nothing, other than that electron-builder exists, and it builds fine on my local machine

**Comment by sgn** [comment]
It may relate to the fact that CI run `xbps-src` as root. I can reproduce in my docker.

**Comment by sgn** [comment]
```diff
diff --git a/srcpkgs/joplin-desktop/template b/srcpkgs/joplin-desktop/template
index 669756d2b5..410abb23be 100644
--- a/srcpkgs/joplin-desktop/template
+++ b/srcpkgs/joplin-desktop/template
@@ -22,7 +22,7 @@ do_build() {
 	# Remove unused modules
 	rm -rf packages/{app-mobile,app-cli,generator-joplin,app-clipper}
 
-	npm install
+	npm install --unsafe-perm
 	cd packages/app-desktop
 	npm run dist
 }
```

**Comment by sgn** [comment]
Actually, this is what I think better:
```diff
diff --git a/srcpkgs/joplin-desktop/template b/srcpkgs/joplin-desktop/template
index 410abb23be..ff200fcd7c 100644
--- a/srcpkgs/joplin-desktop/template
+++ b/srcpkgs/joplin-desktop/template
@@ -22,7 +22,7 @@ do_build() {
 	# Remove unused modules
 	rm -rf packages/{app-mobile,app-cli,generator-joplin,app-clipper}
 
-	npm install --unsafe-perm
+	npm install ${XBPS_ALLOW_CHROOT_BREAKOUT:+--unsafe-perm}
 	cd packages/app-desktop
 	npm run dist
 }
```

**Comment by fosslinux** [comment]
I agree, thank you. I will fix

**Comment by fosslinux** [comment]
@sgn could you check if something similar happens for mattermost-desktop element-desktop and friends? It may need a similar fix.

File: srcpkgs/joplin-cli/template [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+homepage="https://joplinapp.org"
+distfiles="https://github.com/laurent22/joplin/archive/v${version}.tar.gz"
+checksum=55aad4fe50e2da980983a69bc7c0870626064db971550d522e266feb17d38916
+python_version=3
+
+do_build() {
+	# Remove unused modules
+	rm -rf packages/{app-mobile,app-desktop,app-clipper,generator-joplin}
+
+	npm install
```
Comment (sgn) [inline srcpkgs/joplin-cli/template:21]: It need a fix here, too.

**Comment by sgn** [comment]
This could be used to fix cross, but I don't know why npm doesn't pick wrapped `pkg-config`
```diff
diff --git a/srcpkgs/joplin-cli/template b/srcpkgs/joplin-cli/template
index 2f04a10dd0..abfeddbd98 100644
--- a/srcpkgs/joplin-cli/template
+++ b/srcpkgs/joplin-cli/template
@@ -14,6 +14,12 @@ distfiles="https://github.com/laurent22/joplin/archive/v${version}.tar.gz"
 checksum=55aad4fe50e2da980983a69bc7c0870626064db971550d522e266feb17d38916
 python_version=3
 
+if [ "$CROSS_BUILD" ]; then
+	export PKG_CONFIG_SYSROOT_DIR="$XBPS_CROSS_BASE"
+	export PKG_CONFIG_LIBDIR="$XBPS_CROSS_BASE/usr/lib/pkgconfig"
+	export PKG_CONFIG_PATH="$PKG_CONFIG_LIBDIR:$XBPS_CROSS_BASE/usr/share/pkgconfig"
+fi
+
 do_build() {
 	# Remove unused modules
 	rm -rf packages/{app-mobile,app-desktop,app-clipper,generator-joplin}
diff --git a/srcpkgs/joplin-desktop/template b/srcpkgs/joplin-desktop/template
index ff200fcd7c..6fc7ab0faf 100644
--- a/srcpkgs/joplin-desktop/template
+++ b/srcpkgs/joplin-desktop/template
@@ -13,11 +13,18 @@ license="MIT"
 homepage="https://joplinapp.org"
 distfiles="https://github.com/laurent22/joplin/archive/v${version}.tar.gz"
 checksum=55aad4fe50e2da980983a69bc7c0870626064db971550d522e266feb17d38916
+no_generic_pkgconfig_link=yes
 
 export USE_SYSTEM_7ZA="true"
 export USE_SYSTEM_APP_BUILDER="true"
 export USE_SYSTEM_MKSQUASHFS="true"
 
+if [ "$CROSS_BUILD" ]; then
+	export PKG_CONFIG_SYSROOT_DIR="$XBPS_CROSS_BASE"
+	export PKG_CONFIG_LIBDIR="$XBPS_CROSS_BASE/usr/lib/pkgconfig"
+	export PKG_CONFIG_PATH="$PKG_CONFIG_LIBDIR:$XBPS_CROSS_BASE/usr/share/pkgconfig"
+fi
+
 do_build() {
 	# Remove unused modules
 	rm -rf packages/{app-mobile,app-cli,generator-joplin,app-clipper}
```

**Comment by fosslinux** [comment]
Sorry, I didn't test crosss... oops. Will push fixes now.

**Comment by fosslinux** [comment]
I wonder, can we make CI not fail for a cross dependency...

**Comment by fosslinux** [comment]
Patch dosen't work, will investigate further. Seems to be using host cflags but cross CC.

**Comment by fosslinux** [comment]
No idea what's happening here, somehow `-m64` appears out of nowhere, breaking the build. Going to leave joplin-cli as nocross for now.

**Comment by fosslinux** [comment]
WTF is up with i686 CI?

**Comment by fosslinux** [comment]
Ping

**Comment by fosslinux** [comment]
Updated to 1.5.14, i686 issue could be reproduced there, oddly, and is now fixed. Cross joplin-cli is still an issue. Tests enabled on joplin-desktop as they work but not on joplin-cli as it requires a hard-to-package toll which I might or might not come back to (jest, FWIW).

**Comment by fosslinux** [comment]
musl runtime is broken

**Comment by fosslinux** [comment]
musl has now been fixed on both joplin-desktop and joplin-cli. It has been runtime tested extensively on x86_64{,-musl} and is ready for merge.

**Comment by ericonr** [comment]
Segfaults when launched, on msul + wayland. Haven't tried on X.

**Comment by ericonr** [comment]
@fosslinux are you interested in merging this even with the intermittent status of electron on musl?

It looks generally ok.

**Comment by fosslinux** [comment]
Yes I am still interested in getting this merged.

Perhaps I should update the package before merge though... come back in 24 hrs :P

**Comment by fosslinux** [comment]
Ping.

- cross has been fixed
- updated to latest version

**Comment by notramo** [comment]
Why does it segfault on musl Wayland? It's probably not because of Electron as Element Desktop runs fine.

**Comment by ericonr** [comment]
Electron is very fiddly on musl. I had crashes with vscode but not element-desktop, for example.

**Pull request closed without merging**